benchmarks/hash_table/static_reduction_map/param_sweep.cu

```diff
@@ -0,0 +1,102 @@
+ * limitations under the License.
+ */
+
+#include <cuco/static_reduction_map.cuh>
+#include <key_generator.hpp>
+#include <nvbench/nvbench.cuh>
+#include <thrust/device_vector.h>
+#include <thrust/functional.h>
+#include <utils.hpp>
+
+/**
+ * @brief Grid search evaluating backoff delay params for cuco::custom_op
+ */
+template <typename Key,
+          typename Value,
+          nvbench::int32_t BackoffBaseDelay,
+          nvbench::int32_t BackoffMaxDelay>
+void nvbench_cuco_static_reduction_map_custom_op_backoff_delay(
+  nvbench::state& state,
+  nvbench::type_list<Key,
+                     Value,
+                     nvbench::enum_type<BackoffBaseDelay>,
+                     nvbench::enum_type<BackoffMaxDelay>>)
+{
+  using custom_op_type =
+    cuco::custom_op<Value, 0, thrust::plus<Value>, BackoffBaseDelay, BackoffMaxDelay>;
+  using map_type = cuco::static_reduction_map<custom_op_type, Key, Value>;
+
+  auto const num_elems    = state.get_int64("NumInputs");
+  auto const occupancy    = state.get_float64("Occupancy");
+  auto const dist         = state.get_string("Distribution");
+  auto const multiplicity = state.get_int64_or_default("Multiplicity", 8);
+
+  std::vector<Key> h_keys(num_elems);
+  std::vector<Value> h_values(num_elems);
+
+  if (not generate_keys<Key>(dist, h_keys.begin(), h_keys.end(), multiplicity)) {
+    state.skip("Invalid input distribution.");
+    return;
+  }
+
+  // generate uniform random values
+  generate_keys<Value>("UNIFORM", h_values.begin(), h_values.end(), 1);
+
+  // the size of the hash table under a given target occupancy depends on the
+  // number of unique keys in the input
+  std::size_t const unique   = count_unique(h_keys.begin(), h_keys.end());
+  std::size_t const capacity = std::ceil(SDIV(unique, occupancy));
+
+  // alternative occupancy calculation based on the total number of inputs
+  // std::size_t const capacity = num_elems / occupancy;
+
+  thrust::device_vector<Key> d_keys(h_keys);
+  thrust::device_vector<Value> d_values(h_values);
+
+  auto d_pairs_begin =
+    thrust::make_zip_iterator(thrust::make_tuple(d_keys.begin(), d_values.begin()));
+  auto d_pairs_end = d_pairs_begin + num_elems;
+
+  state.add_element_count(num_elems);
+
+  state.exec(nvbench::exec_tag::sync | nvbench::exec_tag::timer,
+             [&](nvbench::launch& launch, auto& timer) {
+               map_type map{capacity, -1};
+
+               timer.start();
+               map.insert(d_pairs_begin, d_pairs_end, launch.get_stream());
+               timer.stop();
+             });
+}
+
+// type parameter dimensions for benchmark
+using key_type_range   = nvbench::type_list<nvbench::int32_t, nvbench::int64_t>;
+using value_type_range = nvbench::type_list<nvbench::int32_t, nvbench::int64_t>;
+using base_delay_range = nvbench::enum_type_list<4, 8, 16, 32, 64, 128, 256, 512>;
+using max_delay_range  = nvbench::enum_type_list<2'048, 4'096, 8'192, 16'384>;
+
+// benchmark setups
+NVBENCH_BENCH_TYPES(
```

Comment:
How much time does it take to run `cuco_static_reduction_map_custom_op_backoff_delay`? Looks like there are totally `2 x 2 x 8 x 4 x 7 = 896` test cases here.

Reply:
Yep, that runs for a looong time. That's why I put it into a separate [binary](https://github.com/sleeepyjack/cuCollections/blob/6d05ebb82938fb862a460e5bbd3dd579b11b7083/benchmarks/CMakeLists.txt#L92). It's basically a parameter grid search. Dunno if this should go into `dev`.